src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs

```diff
@@ -0,0 +1,216 @@
+        {
+            _config = config;
+            _counters = new int[ExponentArraySize][];
+            if (_config.MaxRelativeError < MinRelativeError)
+            {
+                // Ensure that we don't create enormous histograms trying to get overly high precision
+                throw new ArgumentException();
+            }
+            int mantissaBits = (int)Math.Ceiling(Math.Log(1 / _config.MaxRelativeError, 2)) - 1;
+            _mantissaShift = 52 - mantissaBits;
+            _mantissaMax = 1 << mantissaBits;
+            _mantissaMask = _mantissaMax - 1;
+        }
+
+        public override IAggregationStatistics Collect()
+        {
+            int[][] counters;
+            int count;
+            lock (this)
+            {
+                counters = _counters;
+                count = _count;
+                _counters = new int[ExponentArraySize][];
+                _count = 0;
+            }
+            QuantileValue[] quantiles = new QuantileValue[_config.Quantiles.Length];
+            int nextQuantileIndex = 0;
+            if (nextQuantileIndex == _config.Quantiles.Length)
+            {
+                return new HistogramStatistics(quantiles);
+            }
+
+            // Reduce the count if there are any NaN or +/-Infinity values that were logged
+            count -= GetInvalidCount(counters);
+
+            // Consider each bucket to have N entries in it, and each entry has value GetBucketCanonicalValue().
+            // If all these entries were inserted in a sorted array, we are trying to find the value of the entry with
+            // index=target.
+            int target = QuantileToRank(_config.Quantiles[nextQuantileIndex], count);
+
+            // the total number of entries in all buckets iterated so far
+            int cur = 0;
+            foreach (Bucket b in IterateBuckets(counters))
+            {
+                cur += b.Count;
+                while (cur > target)
+                {
+                    quantiles[nextQuantileIndex] = new QuantileValue(
+                        _config.Quantiles[nextQuantileIndex], b.Value);
+                    nextQuantileIndex++;
+                    if (nextQuantileIndex == _config.Quantiles.Length)
+                    {
+                        return new HistogramStatistics(quantiles);
+                    }
+                    target = QuantileToRank(_config.Quantiles[nextQuantileIndex], count);
+                }
+            }
+
+            Debug.Assert(count == 0);
+#if NO_ARRAY_EMPTY_SUPPORT
+            return new HistogramStatistics(EmptyQuantiles);
+#else
+            return new HistogramStatistics(Array.Empty<QuantileValue>());
+#endif
+        }
+
+        private int GetInvalidCount(int[][] counters)
+        {
+            int[] positiveInfAndNan = counters[ExponentArraySize / 2 - 1];
+            int[] negativeInfAndNan = counters[ExponentArraySize - 1];
+            int count = 0;
+            if (positiveInfAndNan != null)
```

Comment:
I'm confused by the nullability annotations here.  If positiveInfAndNan could be null, seems we're missing some question marks in various places.

Reply:
Fixed. It should have been int[]?[]